internal/resource/vault.go

```diff
@@ -0,0 +1,158 @@
+package resource
+
+import (
+	"context"
+	"fmt"
+
+	v1 "github.com/kong/koko/internal/gen/grpc/kong/admin/model/v1"
+	"github.com/kong/koko/internal/model"
+	"github.com/kong/koko/internal/model/json/extension"
+	"github.com/kong/koko/internal/model/json/generator"
+	"github.com/kong/koko/internal/model/json/validation"
+	"github.com/kong/koko/internal/model/json/validation/typedefs"
+)
+
+const (
+	TypeVault model.Type = "vault"
+)
+
+func NewVault() Vault {
+	return Vault{
+		Vault: &v1.Vault{},
+	}
+}
+
+type Vault struct {
+	Vault *v1.Vault
+}
+
+func (r Vault) ID() string {
+	if r.Vault == nil {
+		return ""
+	}
+	return r.Vault.Id
+}
+
+func (r Vault) Type() model.Type {
+	return TypeVault
+}
+
+func (r Vault) Resource() model.Resource {
+	return r.Vault
+}
+
+// SetResource implements the Object.SetResource interface.
+func (r Vault) SetResource(pr model.Resource) error { return model.SetResource(r, pr) }
+
+func (r Vault) Validate(ctx context.Context) error {
+	return validation.Validate(string(TypeVault), r.Vault)
+}
+
+func (r Vault) ProcessDefaults(ctx context.Context) error {
+	if r.Vault == nil {
+		return fmt.Errorf("invalid nil resource")
+	}
+	defaultID(&r.Vault.Id)
+	return nil
+}
+
+func (r Vault) Indexes() []model.Index {
+	return []model.Index{
+		{
+			Name:      "prefix",
+			Type:      model.IndexUnique,
+			Value:     r.Vault.Prefix,
+			FieldName: "prefix",
+		},
+	}
+}
+
+func init() {
+	err := model.RegisterType(TypeVault, &v1.Vault{}, func() model.Object {
+		return NewVault()
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	vaultTypes := []interface{}{
+		"env",
+	}
+	enterpriseVaultTypes := []interface{}{
+		"aws",
+		"gcp",
+		"hcv",
+	}
+
+	envConfigSchema := &generator.Schema{
+		Type: "object",
+		Properties: map[string]*generator.Schema{
+			"prefix": {
+				Type:    "string",
+				Pattern: "^[a-zA-Z_][a-zA-Z0-9_]*$",
+			},
+		},
+		AdditionalProperties: &falsy,
+	}
+	vaultSchema := &generator.Schema{
+		Properties: map[string]*generator.Schema{
+			"id": typedefs.ID,
+			"prefix": {
+				Type:    "string",
+				Pattern: "^[a-z][a-z0-9-]*[a-z0-9]+$",
+				Not: &generator.Schema{
+					AnyOf: []*generator.Schema{
+						{
+							Enum: vaultTypes,
+						},
+						{
+							Enum: enterpriseVaultTypes,
+						},
+					},
+					Description: fmt.Sprintf("must not be any of %v or %v", vaultTypes, enterpriseVaultTypes),
+				},
+			},
+			"name": {
+				Type: "string",
+				Enum: vaultTypes,
+			},
+			"config": {
+				Type: "object",
```

Comment:
Since this object is going to be defined by some rule based on the vault name, this should probably be written like this:

```
			"config": {
				Type:                 "object",
				AdditionalProperties: &truthy,
			},
```